internal/v4/auth.go

```diff
@@ -78,6 +79,32 @@ func (h *Handler) authorize(req *http.Request, acl []string) error {
 	return httpbakery.NewDischargeRequiredError(m, verr)
 }
 
+// checkRequest performs an httpbakery.CheckRequest and then filters out any
+// declared caveat attributes that were not requested.
+func (h *Handler) checkRequest(req *http.Request) (map[string]string, error) {
```

Comment:
Let's make this return a struct:

    type authDeclaredAttrs struct {
        Username string
        Groups []string
    }

That way we know exactly what we've got, and this function
becomes simpler too:

    return &authDeclaredAttrs{
        Username: attrs[usernameAttr],
        Groups: strings.Fields(attrs[groupsAttr]),
    }